File: packages/core/src/marginalia_core/ports/events.py
"""Event publishing ports."""

from __future__ import annotations

from collections.abc import Callable
from typing import Protocol

from marginalia_core.events.models import DomainEvent, EventName

EventHandler = Callable[[DomainEvent], None]


class EventPublisher(Protocol):
    """Publish domain events."""

    def publish(self, event: DomainEvent) -> None:
        """Publish a single event."""
        ...


class EventSubscriber(Protocol):
    """Subscribe to specific event types."""

    def subscribe(self, event_name: EventName, handler: EventHandler) -> None:
        """Register an event handler."""
        ...

File: crates/marginalia-core/src/ports/events.rs
use crate::events::{DomainEvent, EventName};

pub trait EventPublisher {
    fn publish(&self, event: DomainEvent);
}

pub trait EventSubscriber {
    fn subscribe(&mut self, event_name: EventName);
}

impl<T> EventPublisher for &T
where
    T: EventPublisher + ?Sized,
{
    fn publish(&self, event: DomainEvent) {
        (**self).publish(event);
    }
}
